library/core/src/main/java/com/google/android/exoplayer2/audio/MediaCodecAudioRenderer.java

```diff
@@ -168,6 +171,32 @@ public MediaCodecAudioRenderer(
             .build());
   }
 
+
+
+  private final HashMap<String, Integer> AC4_DIALOG_ENHANCEMENT_LEVEL_MAP = new HashMap<String, Integer>() {{
+    put(MediaCodecParameters.DIALOG_ENHANCEMENT_OFF,0);
+    put(MediaCodecParameters.DIALOG_ENHANCEMENT_LEVEL_LOW,3);
+    put(MediaCodecParameters.DIALOG_ENHANCEMENT_LEVEL_MID,6);
+    put(MediaCodecParameters.DIALOG_ENHANCEMENT_LEVEL_HIGH,9);
+  }};
+
+  private void setDialogEnhancement() {
+      if (audioCodecInfo != null && audioCodecInfo.name.contains("ac4")) {
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
+          String dialogEnhancementLevel = (String) mediaCodecParameters.getOrDefault(MediaCodecParameters.KEY_DIALOG_ENHANCEMENT, "off");
+          int    gain = AC4_DIALOG_ENHANCEMENT_LEVEL_MAP.getOrDefault(dialogEnhancementLevel,0);
+          Log.e(TAG, "setDialogEnhancement on AC4 Codec to " + gain + " dB (" + dialogEnhancementLevel + ")");
+          Bundle bundle = new Bundle(1);
+          bundle.putInt("vendor.dolby.dialog-enhancement-gain.value", gain);
+          super.getCodec().setParameters(bundle);
```

Comment:
Do you know what happens if you call `setParameters("key1", value1)` and then later call `setParameters("key2", value2)`? Is the end result that the codec has both parameters set, or does the second call clear out the value passed by the first?

I think for this to work, you need the end result to be that the codec has both parameters set. Else we'll have to find a slightly different way to do things.

Reply:
I have verified that I could set aac-drc-boost-level and aac-drc-cut-level  in 2 consecutive calls to codec.setParameters(bundle)  and that each were applied at some point  (the 2 set could be applied on different output buffers because the codec may be decoding a few buffers in between addressing the 2 setParameter calls. 


...
05-31 17:35:45.018 15168 15228 D CODEC.SETPARAMETERS: set cut-level to 16
05-31 17:35:45.023 15168 15228 D CODEC.SETPARAMETERS:  ... boost-level to 24
**05-31 17:35:45.038 15168 15246 D CCodecBuffers:   int32_t aac-drc-boost-level = 24
05-31 17:35:45.038 15168 15246 D CCodecBuffers:   int32_t aac-drc-cut-level = 16**
...
05-31 17:35:45.092 15168 15228 D CODEC.SETPARAMETERS: set cut-level to 4
05-31 17:35:45.095 15168 15228 D CODEC.SETPARAMETERS:  ... boost-level to 12
**05-31 17:35:45.193 15168 15246 D CCodecBuffers:   int32_t aac-drc-boost-level = 12
05-31 17:35:45.193 15168 15246 D CCodecBuffers:   int32_t aac-drc-cut-level = 4**
...